libs/community/langchain_community/vectorstores/aperturedb.py

```diff
@@ -0,0 +1,446 @@
+                }
+            }
+        ]
+        r, b = self.connection.query(find_ds_query)
+        assert self.connection.last_query_ok(), r
+        n_entities = (
+            len(r[0]["FindDescriptorSet"]["entities"])
+            if "entities" in r[0]["FindDescriptorSet"]
+            else 0
+        )
+        assert n_entities <= 1, "Multiple descriptor sets with the same name"
+
+        if n_entities == 1:  # Descriptor set exists already
+            e = r[0]["FindDescriptorSet"]["entities"][0]
+            self.logger.info(f"Descriptor set {descriptor_set} already exists")
+
+            engines = e["_engines"]
+            assert len(engines) == 1, "Only one engine is supported"
+
+            if self.engine is None:
+                self.engine = engines[0]
+            elif self.engine != engines[0]:
+                self.logger.error(f"Engine mismatch: {self.engine} != {engines[0]}")
+
+            metrics = e["_metrics"]
+            assert len(metrics) == 1, "Only one metric is supported"
+            if self.metric is None:
+                self.metric = metrics[0]
+            elif self.metric != metrics[0]:
+                self.logger.error(f"Metric mismatch: {self.metric} != {metrics[0]}")
+
+            dimensions = e["_dimensions"]
+            if self.dimensions is None:
+                self.dimensions = dimensions
+            elif self.dimensions != dimensions:
+                self.logger.error(
+                    f"Dimensions mismatch: {self.dimensions} != {dimensions}"
+                )
+
+            self.properties = {
+                k[len(PROPERTY_PREFIX) :]: v
+                for k, v in e.items()
+                if k.startswith(PROPERTY_PREFIX)
+            }
+
+        else:
+            self.logger.info(
+                f"Descriptor set {descriptor_set} does not exist. Creating it"
+            )
+            if self.engine is None:
+                self.engine = ENGINE
+            if self.metric is None:
+                self.metric = METRIC
+            if self.dimensions is None:
+                self.dimensions = len(self.embedding_function.embed_query("test"))
+
+            properties = (
+                {PROPERTY_PREFIX + k: v for k, v in self.properties.items()}
+                if self.properties is not None
+                else None
+            )
+
+            self.utils.add_descriptorset(
+                name=descriptor_set,
+                dim=self.dimensions,
+                engine=self.engine,
+                metric=self.metric,
+                properties=properties,
+            )
+
+            # Create indexes
+            self.utils.create_entity_index("_Descriptor", "_create_txn")
+            self.utils.create_entity_index("_DescriptorSet", "_name")
+            self.utils.create_entity_index("_Descriptor", UNIQUEID_PROPERTY)
+
+    @override
+    def add_documents(self, documents: List[Document], **kwargs: Any) -> List[str]:
```

Comment:
You can implement `upsert` now instead of `add_documents` or `add_texts`

Reply:
OK.  I saw that `upsert` got added, but I chose to implement it using `add_documents`.  Now that `add_documents` and `add_texts` have default implementations that use `upsert`, I could reverse that.

Reply:
As noted above, I have recoded to remove `add_documents`and `add_texts` in favour of `upsert`.